### ISSUE TITLE
Rare error that may cause lsp to crash

### ISSUE THREAD

**Issue by renkun-ken** [REditorSupport/languageserver#127]
When I open `R/utils.R`, the following error occurs and lsp crashed. But I cannot reproduce this error at the moment.

```
received:  Content-Length: 9352
error handling json:  {"message":"lexical error: invalid char in json text.\n             } else if (tag 464288n","call":{}}
received:  \\\\ifelse\", \"USERMACRO\")) {\n            \"\"\n        } else if (is.character(item)) {\n            trimws(item)\n        } else if (length(item)) {\n            convert_doc_to_markdown(item)\n        }\n    }))\n}\n\nconvert_doc_string <- function(doc) {\n    paste0(convert_doc_to_markdown(doc), collapse = \" \")\n}\n\nxdoc_find_enclosing_scopes <- function(x, line, col, top = FALSE) {\n    if (top) {\n        xpath <- glue(\"/exprlist | //expr[(@line1 < {line} or (@line1 = {line} and @col1 <= {col})) and\n                (@line2 > {line} or (@line2 = {line} and @col2 >= {col}))]\")\n    } else {\n        xpath <- glue(\"//expr[(@line1 < {line} or (@line1 = {line} and @col1 <= {col})) and\n                (@line2 > {line} or (@line2 = {line} and @col2 >= {col}))]\")\n    }\n    xml_find_all(x, xpath)\n}\n\nxdoc_find_token <- function(x, line, col) {\n    xpath <- glue(\"//*[not(*)]\n    [(@line1 < {line} or (@line1 = {line} and @col1 <= {col})) and\n                (@line2 > {line} or (@line2 = {line} and @col2 >= {col}))]\")\n    xml_find_first(x, xpath)\n}\n\nxml_single_quote <- function(x) {\n    x <- gsub(\"'\", \"&apos;\", x, fixed = TRUE)\n    x\n}\n"}}}Content-Length: 161
Error in self$read_header() : Unexpected non-empty line

[]
exiting
[Info  - 12:13:07 AM] Connection to server got closed. Server will restart.
[Error - 12:13:07 AM] Request textDocument/documentSymbol failed.
Error: Connection got disposed.
	at Object.dispose (/Users/ken/.vscode/extensions/reditorsupport.r-lsp-0.1.4/node_modules/vscode-jsonrpc/lib/main.js:876:25)
	at Object.dispose (/Users/ken/.vscode/extensions/reditorsupport.r-lsp-0.1.4/node_modules/vscode-languageclient/lib/client.js:71:35)
	at LanguageClient.handleConnectionClosed (/Users/ken/.vscode/extensions/reditorsupport.r-lsp-0.1.4/node_modules/vscode-languageclient/lib/client.js:2153:42)
	at LanguageClient.handleConnectionClosed (/Users/ken/.vscode/extensions/reditorsupport.r-lsp-0.1.4/node_modules/vscode-languageclient/lib/main.js:151:15)
	at closeHandler (/Users/ken/.vscode/extensions/reditorsupport.r-lsp-0.1.4/node_modules/vscode-languageclient/lib/client.js:2140:18)
	at CallbackList.invoke (/Users/ken/.vscode/extensions/reditorsupport.r-lsp-0.1.4/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/Users/ken/.vscode/extensions/reditorsupport.r-lsp-0.1.4/node_modules/vscode-jsonrpc/lib/events.js:120:36)
	at closeHandler (/Users/ken/.vscode/extensions/reditorsupport.r-lsp-0.1.4/node_modules/vscode-jsonrpc/lib/main.js:226:26)
	at CallbackList.invoke (/Users/ken/.vscode/extensions/reditorsupport.r-lsp-0.1.4/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/Users/ken/.vscode/extensions/reditorsupport.r-lsp-0.1.4/node_modules/vscode-jsonrpc/lib/events.js:120:36)
	at StreamMessageReader.fireClose (/Users/ken/.vscode/extensions/reditorsupport.r-lsp-0.1.4/node_modules/vscode-jsonrpc/lib/messageReader.js:111:27)
	at Socket.<anonymous> (/Users/ken/.vscode/extensions/reditorsupport.r-lsp-0.1.4/node_modules/vscode-jsonrpc/lib/messageReader.js:151:46)
	at Socket.emit (events.js:205:15)
	at Pipe.<anonymous> (net.js:586:12)
```

**Comment by randy3k** [comment]
I also saw it occasionally on CI tests. I thought I have fixed it already, let me take a look again.

**Comment by randy3k** [comment]
Do you know whese `             } else if (tag 464288n","call":{}}` comes from? It doesn't seem to be from language server codebase.

**Comment by renkun-ken** [comment]
It looks like this occurs when the language server is not ready or the request header is not complete but the request is processed anyway only to find it is incomplete.

**Comment by renkun-ken** [comment]
In stdio mode, when the script is very long (e.g. 2000+ lines of code), it is easy to crash the language server. The same code, same editing, work well in TCP mode. And the error above is in stdio mode too.

Maybe a different error here but crashes with segfault:

```
[Error - 9:14:46 AM] Connection to server is erroring. Shutting down server.
[Error - 9:14:46 AM] Connection to server is erroring. Shutting down server.

 *** caught segfault ***
address 0x114148fff, cause 'memory not mapped'

Traceback:
 1: stdin_read_char(n)
 2: self$read_char(nbytes)
 3: self$read_content(nbytes)
 4: self$fetch(blocking = FALSE)
 5: doTryCatch(return(expr), name, parentenv, handler)
 6: tryCatchOne(expr, names, parentenv, handlers[[1L]])
 7: tryCatchList(expr, classes, parentenv, handlers)
 8: tryCatch(expr, error = function(e) {    call <- conditionCall(e)    if (!is.null(call)) {        if (identical(call[[1L]], quote(doTryCatch)))             call <- sys.call(-4L)        dcall <- deparse(call)[1L]        prefix <- paste("Error in", dcall, ": ")        LONG <- 75L        sm <- strsplit(conditionMessage(e), "\n")[[1L]]        w <- 14L + nchar(dcall, type = "w") + nchar(sm[1L], type = "w")        if (is.na(w))             w <- 14L + nchar(dcall, type = "b") + nchar(sm[1L],                 type = "b")        if (w > LONG)             prefix <- paste0(prefix, "\n  ")    }    else prefix <- "Error : "    msg <- paste0(prefix, conditionMessage(e), "\n")    .Internal(seterrmessage(msg[1L]))    if (!silent && isTRUE(getOption("show.error.messages"))) {        cat(msg, file = outFile)        .Internal(printDeferredWarnings())    }    invisible(structure(msg, class = "try-error", condition = e))})
 9: try({    if (isTRUE(self$exit_flag)) {        logger$info("exiting")        break    }    self$process_events()    data <- self$fetch(blocking = FALSE)    if (is.null(data)) {        Sys.sleep(0.1)        next    }    self$handle_raw(data)}, silent = TRUE)
10: langserver$run()
11: languageserver::run(debug = TRUE)
An irrecoverable exception occurred. R is aborting now ...
```

I'll come up with non-sense and correct long script to reproduce this.

**Comment by randy3k** [comment]
I know what's happening here. It is a buffer overflow error. Hopefully will fix in #140.
